tests/60app-services/01as-create.pl

```diff
@@ -27,6 +27,40 @@
       });
    };
 
+test "AS can get or create user and return an access_token",
+   requires => [ $main::AS_USER[0], $main::API_CLIENTS[0] ],
+
+   do => sub {
+      my ( $as_user, $http ) = @_;
+
+      do_request_json_for( $as_user,
+         method => "POST",
+         uri    => "/unstable/createUser",
+
+         content => {
+            localpart        => "user_localpart",
+            displayname      => "user_displayname",
+            duration_seconds => 200,
+         },
+      )->then( sub {
+         my ( $body ) = @_;
+
+         assert_json_keys( $body, qw( access_token user_id  home_server ));
+
+         my $user =User( $http , $body->{user_id}, "", $body->{access_token}, undef, undef, undef, [], undef );
+
+         do_request_json_for( $user,
+            method => "GET",
+            uri    => "/r0/profile/:user_id/displayname",
+         )})->then(sub {
```

Comment:
This doesn't need to be nested, as `->then( ... )` chains things correctly. For example:

```perl
foo()
->then( sub {
   bar();
})->then( sub {
   my ( $bar_result ) = @_;

   ...
});
```

Does the right thing, i.e. `$bar_result` will be the result of the `bar()` invocation, assuming it returns a promise/deferred.

(In perl, the last expression in a `sub { ... }` is used as the return value.)

This is really nice as it means you don't end up with tonnes of nesting, and generally makes it easier to read.

---

You also want a space between the parenthesis and `sub`, lest you make leo sad.